RFC-0016-Masked-reductions-and-normalizations.md

```diff
@@ -0,0 +1,259 @@
+* min
+* dist
+* logsumexp
+* mean
+* median
+* nanmedian
+* mode
+* norm
+* nansum
+* prod
+* quantile
+* nanquantile
+* std
+* std_mean
+* sum
+* unique
+* unique_consecutive
+* var
+* var_mean
+* count_nonzero
+
+Sparse COO ([docs](https://pytorch.org/docs/master/sparse.html#torch-functions-specific-to-sparse-tensors))
+
+
+
+* sum
+
+Sparse CSR: None
+
+**Existing normalization operators**
+
+These are a bit more difficult to list exhaustively, because we don't have an explicit category for them. A normalization operator applies a reduction to a Tensor slice and uses the result to normalize the slice.
+
+Dense ([functionals](https://pytorch.org/docs/master/nn.functional.html))
+
+
+
+* softmax
+* log_softmax
+* batch_norm
+* group_norm
+* layer_norm
+* instance_norm
+* local_response_norm
+* normalize
+* cross_entropy
+* nll_loss
+
+Sparse COO ([docs](https://pytorch.org/docs/master/sparse.html#torch-functions-specific-to-sparse-tensors))
+
+
+
+* softmax
+* log_softmax
+* norm / native_norm
+
+Sparse CSR: None
+
+**Operator constraints and general signature - Reductions**
+
+**Input types:** A mask is a boolean tensor. It accompanies a dense Tensor of the same shape. If an entry is True the corresponding element at the same index in the paired dense Tensor is a "valid" value. If it is False it is not. "valid" here means that this value is meant to be included in the computation and otherwise is meant to be ignored. This matches the semantics of masked_scatter, masked_select and masked_fill. 
+
+**Fully masked rows:** If a slice (e.g. row) is fully masked out there is no guarantee the corresponding return values are filled with any specific value such as the operation's identity value. However, given a sparse input with a row entirely zero and masked out the result is likely to be zero to maximize memory savings.
+
+**Operator coverage:** As a first step we should implement the basics: sum, prod, mean, amin, amax.
+
+**Semantics:** Indeed the best way to describe the behavior is to implement it. Please note that this is only meant to describe semantics and is not an actual implementation. It also implies the gradients for these operations. Whether an input or mask is backed by a dense or compressed storage layout has no influence on the semantics of these operations. More concretely, unspecified values for sparse layouts are always zero in this context. If the mask is None the behavior of the operator is equivalent to the unmasked variant.
+
+**Input layouts:** If an input is passed a sparse Tensor, it will return a sparse output. At first the layout of the mask and input must match. If the input is a sparse Tensor, the layout of the mask must be too. This is to ease kernel development and might quickly change, if there is user demand. However, the user can always convert a mask to sparse or dense before invocation.
+
+
+```
+def masked_sum(input, dim, keepdim, dtype, mask):
+    return torch.sum(input * mask, dim, keepdim, dtype=dtype)
```

Comment:
NOTE: This actually requires masked_fill, because we have no guarantee that any masked values of input are valid.
```
>>> torch.tensor([float('inf')]) * torch.tensor([False])
tensor([nan])
>>> torch.tensor([float('inf')]).masked_fill(~torch.tensor([False]), 0)
tensor([0.])
```